### PR TITLE
ipam/crd: Fix may allocated IPs which have been marked as ready-for-release

### LINKED ISSUE #32531: In the iPam mode of CRD, the pod may allocate an IP that has already been released
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When using IPAM in CRD mode, I encountered an issue where the IP address already assigned to the pod was marked as released. After analysis, it was found that there are bugs in the current version of Cilium.

The steps and related codes to reproduce the issue are as follows:
1. The operator triggered 'PrepareIPRelease()' and marked an IP in 'CiliumNode. status. release.ips' as 'marked-for-release'. [node.go:L789](https://github.com/cilium/cilium/blob/73b3704bb96bbd90ba76992d9ad978d87e402716/pkg/ipam/node.go#L789)
2. The agent received a new 'CiliumNode' update event and triggered the 'updateLocalNodeResource()` function. [allocator.go#L196](https://github.com/cilium/cilium/blob/73b3704bb96bbd90ba76992d9ad978d87e402716/pkg/ipam/allocator.go#L196)
3. Concurrent with 2., create a new pod and allocate a new IP.
4. Finally, the newly allocated IP for the third part will be marked as' ready-for-release 'in the 2. step, and the IP address release will be completed.  The IP address of the Pod that has already been running was released..

The detailed code related to the problem is as follows:
https://github.com/cilium/cilium/blob/73b3704bb96bbd90ba76992d9ad978d87e402716/pkg/ipam/crd.go#L428-L443

### Cilium Version

1.14.6

### Kernel Version

5.15.54-051554-generic

### Kubernetes Version

1.24.4

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by weiwei02**
Fix the problem that CRD IPAM may apply for IPs that have been marked as ready-for-release.

Fixes: #32531

Due to the sequential locking requirements of the nodeStore and allocator objects, I completed the statistics of IPs to be released within the scope of nodeStore.mutex by using a map. This avoids the problem of IP addresses being allocated to newly created Pods during the unlocked period.

Fixes: #32531 

```release-note
Fix IPAM of CRD mode may apply for IPs that have been marked as ready-for-release.
```

### CODE REVIEW

**Comment by joestringer** [comment]
Hi, thanks for the submission. Do you think it is possible to come up with a unit test to validate the change in functionality?

**Comment by christarazi** [comment]
@hemanthmalla Could you take a look here since you were the most recent codeowner that modified this logic?

**Comment by hemanthmalla** [comment]
@weiwei02 Can you describe the race condition in a little more detail ?

Are we trying to fix the window between allocator unlock and node lock ?

https://github.com/cilium/cilium/blob/04a5d7f559a70b0143e8cb74281bd716f3828c5c/pkg/ipam/crd.go#L434-L435

**Comment by weiwei02** [comment]
> Hi, thanks for the submission. Do you think it is possible to come up with a unit test to validate the change in functionality?
@joestringer Thanks for your advice. I have thought about this question, but unit tests may not be able to describe the data security issues caused by the gap of locks, so I did not add any additional unit tests.

**Comment by weiwei02** [comment]
> @weiwei02 Can you describe the race condition in a little more detail ?@weiwei02你能更详细地描述一下比赛情况吗？
> 
> Are we trying to fix the window between allocator unlock and node lock ?我们是否试图修复分配器解锁和节点锁定之间的窗口？
> 
> https://github.com/cilium/cilium/blob/04a5d7f559a70b0143e8cb74281bd716f3828c5c/pkg/ipam/crd.go#L434-L435

@hemanthmalla Yes, I have described the issue I found in the actual operating environment in detail in #32531. In the updateLocalNodeResource() function, if the lock of nodeStore and the lock of allocator are released simultaneously, during the gap time, the IP address might be allocated by the allocator.

**Comment by ldelossa** [comment]
@weiwei02 please fix the failing lint in `BPF Checks / Check Patch`. Looks like the commit message is too long

File: pkg/ipam/crd.go [hunk 1]
```diff
@@ -418,29 +417,31 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
...
-		if allocator == nil {
+		if _, ok := markedToReleaseIPMap[ip]; !ok {
 			continue
 		}
+		releaseUpstreamSyncNeeded = true
+	}
+	n.mutex.Unlock()
 
+	for ip, allocator := range markedToReleaseIPMap {
 		// Some functions like crdAllocator.Allocate() acquire lock on allocator first and then on nodeStore.
```
Comment (doniacld) [inline pkg/ipam/crd.go]: I believe this comment should live above with `n.mutex.Unlock()`.

File: pkg/ipam/crd.go [hunk 2]
```diff
@@ -418,29 +417,31 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
...
 
-		if ok {
+		if _, ok := allocator.allocated[ip]; ok {
 			// IP still in use, update the operator to stop releasing the IP.
 			n.ownNode.Status.IPAM.ReleaseIPs[ip] = ipamOption.IPAMDoNotRelease
 		} else {
 			n.ownNode.Status.IPAM.ReleaseIPs[ip] = ipamOption.IPAMReadyForRelease
 		}
-		releaseUpstreamSyncNeeded = true
+		n.mutex.Unlock()
```
Comment (doniacld) [inline pkg/ipam/crd.go:446]: Add a comment to explain why we are not deferring it anymore: 
```suggestion
                 // unlock node mutex before the allocator mutex to avoid IP addresses being allocated to newly created pods during the unlocked period
		n.mutex.Unlock()
```

File: pkg/ipam/crd.go [hunk 3]
```diff
@@ -337,14 +337,14 @@ func (n *nodeStore) deleteLocalNodeResource() {
...
 	n.mutex.Lock()
-	defer n.mutex.Unlock()
 
 	if n.conf.IPAMMode() == ipamOption.IPAMENI {
 		if err := configureENIDevices(n.ownNode, node, n.mtuConfig); err != nil {
 			log.WithError(err).Errorf("Failed to update routes and rules for ENIs")
 		}
 	}
 
+	var markedToReleaseIPMap = make(map[string]*crdAllocator)
```
Comment (doniacld) [inline pkg/ipam/crd.go]: I would rename the map. 
```suggestion
	var ipsMarkedToRelease = make(map[string]*crdAllocator)
```

**Review by doniacld — CHANGES_REQUESTED** [review]
Thanks for your PR @weiwei02 !
 
I would rephrase the commit into something like that `ipam/crd: Fix useless IP allocation that was marked `ready-for-release` and add some context in the description from the issue you created (which is well explained 😉).

File: pkg/ipam/crd.go (hunk 1 above)
Comment (weiwei02) [reply pkg/ipam/crd.go]: thanks, i would adjust the position of the comments

**Comment by doniacld** [comment]
There is one last nit, your commit description is missing the `#` before the issue number to get the link. 
Otherwise LGTM! Approving it.

**Review by doniacld — APPROVED** [review]

**Comment by julianwiedmann** [comment]
Please address the [checkpatch errors](https://github.com/cilium/cilium/actions/runs/10939478880/job/30390860526?pr=34489) in your commit message. In particular the author sign-off.

**Comment by aanm** [comment]
@weiwei02 ping?

**Comment by dylandreimerink** [comment]
@weiwei02 Our CI is not happy with the commit message, could you look into this?
